doc/pipeline_details.rst

```diff
@@ -0,0 +1,31 @@
+.. _pipeline-details:
+
+Pipeline Details
+================
+
+Ibex has a 2 stage pipeline, the 2 stages are:
+
+Instruction Fetch (IF)
+  Fetches instructions from memory via a prefetch buffer, capable of fetching 1 instruction per cycle if the instruction side memory system allows. See :ref:`instruction-fetch` for details.
+
+Instruction Decode and Execute (ID/EX)
+  Decodes fetched instruction and immediately executes it, register read and write all occurs in this stage. Multi-cycle instructions will stall this stage until they are complete See :ref:`instruction-decode-execute` for details.
+
+All instructions required two cycles minimum to pass down the pipeline. One cycle in the IF stage and one in the ID/EX stage. Not all instructions can complete in the ID/EX stage in one cycle so will stall there until they complete. This means the maximum CPI (clocks per instruction) Ibex can achieve is 1 when multi-cycle instructions aren't used. See Multi and Single Cycle Instructions below for the details.
+
+Multi and Single Cycle Instructions
+-----------------------------------
+
+All of the 'Integer Computational Instructions' as defined in the RISC-V RV32I Base Integer Instruction set are single cycle. The CSR access instructions defined in 'Zicsr' of the RISC-V specification are also single cycle. All other instructions are multi-cycle. Below are descriptions of the multi-cycle operations for various categories of instructions.
+
+Load/Store operations
+  Both loads and stores stall for at least one cycle to await a response. For loads this response is the load data (which is written directly to the register file the same cycle it is received). For stores this is whether an error was seen or not. The longer the data side memory interface takes to receive a response the longer loads and stores will stall
+
+Multiplication and Division
+ The algorithms used for these take multiple cycles to complete. Multiplication and division operation will stall until their results are available (more detail in :ref:`mult-div`).
+
+Jump
+ Any jump will stall for at least one cycle. This is to flush the prefetch buffer and IF stage and start requesting instructions from a different PC (the new PC request will appear on the instruction side memory interface the same cycle the jump instruction enters ID/EX). The longer the instruction side memory interface takes to receive data the longer the jump will stall.
+
+Branch
+ Any branch will stall for one cycle, any taken branch will stall for at least two cycles. In the first cycle the ALU is used to compute the branch condition. If taken the branch then proceeds as a Jump does above (IF stage and prefetch buffer flushed, ALU generates target, stall for at least one cycle to get new instruction).
```

Comment:
I am not sure if the the description of the jump and branch is correct or if I am a bit confused with what a stall actually is. Looking a the decoder and ID stage sources, I can see that jumps take at least two cycles in ID/EX (calc jump target + issue new fetch in IF, calc and store PC) where as branches take one ore two cycles in ID/EX (eval branch condition, calc jump target + issue new fetch in IF). But if the branch is not taken, the pipeline is not stalled. In both cases (jump, branch taken), the effective number of cycles then depends on what happens in the IF stage, i.e., how long it takes to serve the fetch. I would suggest to reformulate the descriptions along these lines, i.e., from a perspective of the ID/EX stage.